File: WdtcCore/src/main/java/org/wdt/wdtc/download/fabric/FabricInstallTask.java
package org.wdt.wdtc.download.fabric;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import org.wdt.utils.dependency.DependencyDownload;
import org.wdt.wdtc.download.DownloadTask;
import org.wdt.wdtc.download.infterface.DownloadSource;
import org.wdt.wdtc.download.infterface.InstallTask;
import org.wdt.wdtc.game.GameVersionJsonObject;
import org.wdt.wdtc.game.Launcher;
import org.wdt.wdtc.game.LibraryObject;
import org.wdt.wdtc.utils.gson.JSONArray;
import org.wdt.wdtc.utils.gson.JSONObject;
import org.wdt.wdtc.utils.gson.JSONUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FabricInstallTask extends FabricDonwloadInfo implements InstallTask {
    private final DownloadSource source;



    public FabricInstallTask(Launcher launcher, String FabricVersionNumber) {
        super(launcher, FabricVersionNumber);
        this.source = Launcher.getDownloadSource();

    }



    @Override
    public void execute() throws IOException {
        GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
        List<LibraryObject> libraryObjectList = VersionJsonObject.getLibraries();
        JSONObject FabricVersionJsonObject = getFabricVersionJsonObject();
        JSONArray FabricLibraryList = FabricVersionJsonObject.getJSONArray("libraries");
        for (int i = 0; i < FabricLibraryList.size(); i++) {
            JSONObject object = FabricLibraryList.getJSONObject(i);
            DependencyDownload dependency = new DependencyDownload(object.getString("name"));
            dependency.setDefaultUrl(source.getFabricLibraryUrl());
            dependency.setDownloadPath(launcher.getGameLibraryPath());
            libraryObjectList.add(LibraryObject.getLibraryObject(dependency, object.getString("url")));
        }
        VersionJsonObject.setLibraries(libraryObjectList);
        VersionJsonObject.setMainClass(FabricVersionJsonObject.getString("mainClass"));
        GameVersionJsonObject.Arguments arguments = VersionJsonObject.getArguments();
        JSONObject Arguments = FabricVersionJsonObject.getJSONObject("arguments");
        JsonArray JvmList = arguments.getJvmList();
        JvmList.addAll(Arguments.getJSONArray("jvm").getJsonArrays());
        arguments.setJvmList(JvmList);
        JsonArray GameList = arguments.getGameList();
        GameList.addAll(Arguments.getJSONArray("game").getJsonArrays());
        arguments.setGameList(GameList);
        VersionJsonObject.setArguments(arguments);
        VersionJsonObject.setId(launcher.getVersionNumber() + "-fabric-" + FabricVersionNumber);
        launcher.PutToVersionJson(VersionJsonObject);
    }


    @Override
    public void setPatches() throws IOException {
        GameVersionJsonObject Object = launcher.getGameVersionJsonObject();
        List<JsonObject> ObjectList = new ArrayList<>();
        ObjectList.add(JSONUtils.getJsonObject(launcher.getVersionJson()));
        ObjectList.add(JSONUtils.getJsonObject(getFabricVersionJson()));
        Object.setJsonObject(ObjectList);
        launcher.PutToVersionJson(Object);
    }

    @Override
    public void AfterDownloadTask() throws IOException {
        if (getAPIDownloadTaskNoNull()) {
            getAPIDownloadTask().DownloadFabricAPI();
        }
    }

    @Override
    public void BeforInstallTask() {
        DownloadTask.StartDownloadTask(String.format(getFabricVersionFileUrl(), launcher.getVersionNumber(), FabricVersionNumber), getFabricVersionJson());
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/download/fabric/FabricInstallTask.kt
package org.wdt.wdtc.core.impl.download.fabric

import org.wdt.utils.gson.readFileToJsonObject
import org.wdt.wdtc.core.openapi.download.interfaces.ModInstallTaskInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.game.Version
import org.wdt.wdtc.core.openapi.utils.STRING_SPACE
import org.wdt.wdtc.core.openapi.utils.cleanStrInString
import org.wdt.wdtc.core.openapi.utils.noNull
import org.wdt.wdtc.core.openapi.utils.startDownloadTask

class FabricInstallTask(version: Version, fabricVersionNumber: String) :
	FabricDonwloadInfo(version, fabricVersionNumber), ModInstallTaskInterface {
	
	constructor(version: Version, versionsJsonObjectInterface: VersionsJsonObjectInterface) : this(
		version, versionsJsonObjectInterface.versionNumber.noNull()
	)
	
	override fun overwriteVersionJson() {
		fabricProfileJsonObject.also { fabirc ->
			version.run {
				gameVersionJsonObject.apply {
					id = "${versionNumber}-fabric-$modVersion"
					mainClass = fabirc.mainClass
					arguments = arguments.apply {
						jvmList = jvmList.also { game ->
							fabirc.arguments.jvmList?.onEach {
								game!!.add(it.asString.cleanStrInString(STRING_SPACE))
							}
						}
						gameList = gameList.also { game ->
							fabirc.arguments.gameList?.onEach {
								game!!.add(it.asString)
							}
						}
					}
					libraries.addAll(fabirc.libraries.map {
						it.toLibraryObject()
					})
				}.also {
					it.putToVersionJson()
				}
			}
		}
	}
	
	override fun writeVersionJsonPatches() {
		version.run {
			gameVersionJsonObject.apply {
				patches = listOf(
					version.versionJson.readFileToJsonObject(), fabricVersionJson.readFileToJsonObject()
				)
			}.putToVersionJson()
		}
	}
	
	override suspend fun afterDownloadTask() {
		if (isHasApiDownloadTask) {
			apiDownloadTask.noNull().startDownloadFabricAPI()
		}
	}
	
	override suspend fun beforInstallTask() {
		startDownloadTask(
			fabricVersionFileUrl.format(version.versionNumber, modVersion), fabricVersionJson
		)
	}
}